src/Sarif.Viewer.VisualStudio/Tags/SarifLocationTagger.cs

```diff
@@ -0,0 +1,548 @@
+                        // are out of sync which should never happen and is bad.
+                        RunIdToSarifTags[sarifTag.RunId].Remove(sarifTag);
+
+                        sarifTag.PropertyChanged -= this.SarifTagPropertyChanged;
+                        sarifTag.Dispose();
+                    }
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public void RemoveTagsForRun(int runId)
+        {
+            List<SarifLocationTag> tagsToRemove;
+
+            using (TagListLock.EnterReadLock())
+            {
+                if (!RunIdToSarifTags.TryGetValue(runId, out List<SarifLocationTag> sarifTagsForRun) ||
+                    sarifTagsForRun.Count == 0)
+                {
+                    return;
+                }
+
+                // Copy so we can update (which can make outgoing callse) outside of lock
+                tagsToRemove = new List<SarifLocationTag>(sarifTagsForRun);
+            }
+
+            using (this.Update())
+            {
+                foreach (SarifLocationTag tagToRemove in tagsToRemove)
+                {
+                    this.RemoveTag(tagToRemove);
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public IEnumerable<ITagSpan<TextMarkerTag>> GetTags(NormalizedSnapshotSpanCollection spans)
+        {
+            if (spans.Count == 0)
+            {
+                yield break;
+            }
+
+            SarifLocationTag[] possibleTags = null;
+            using (TagListLock.EnterReadLock())
+            {
+                if (this.sarifTags == null)
+                {
+                    yield break;
+                }
+
+                possibleTags = new SarifLocationTag[this.sarifTags.Count];
+                this.sarifTags.CopyTo(possibleTags, 0);
+            }
+
+            if (possibleTags == null)
+            {
+                yield break;
+            }
+
+            foreach (var span in spans)
+            {
+                foreach (var possibleTag in possibleTags.Where((possibleTag) => possibleTag.DocumentPersistentSpan.Span != null))
+                {
+                    SnapshotSpan possibleTagSnapshotSpan = possibleTag.DocumentPersistentSpan.Span.GetSpan(span.Snapshot);
+                    if (span.IntersectsWith(possibleTagSnapshotSpan))
+                    {
+                        yield return new TagSpan<TextMarkerTag>(possibleTagSnapshotSpan, possibleTag.Tag);
+                    }
+                }
+            }
+        }
+
+        /// <inheritdoc/>
+        public IDisposable Update()
+        {
+            return new BatchUpdate(this);
+        }
+
+        protected virtual void Dispose(bool disposing)
+        {
+            if (disposed)
+            {
+                return;
+            }
+
+            disposed = true;
+
+            if (disposing)
+            {
+                using (TagListLock.EnterReadLock())
```

Comment:
>TagListLock [](start = 23, length = 11)

do these locks need to be disposed?

Reply:
In theory, but they are static so we'd need to do it only package unload. I'll take a look.

---
In reply to: [481470693](https://github.com/microsoft/sarif-visualstudio-extension/pull/191#discussion_r481470693) [](ancestors = 481470693)